usr/src/uts/common/fs/zfs/arc.c

```diff
@@ -6709,18 +6709,16 @@ l2arc_evict(l2arc_dev_t *dev, uint64_t distance, boolean_t all)
 			goto top;
 		}
 
-		if (HDR_L2_WRITE_HEAD(hdr)) {
-			/*
-			 * We hit a write head node.  Leave it for
-			 * l2arc_write_done().
-			 */
-			list_remove(buflist, hdr);
-			mutex_exit(hash_lock);
-			continue;
-		}
+		/*
+		 * A header can't be on this list if it doesn't have L2 header.
+		 */
+		ASSERT(HDR_HAS_L2HDR(hdr));
 
-		if (!all && HDR_HAS_L2HDR(hdr) &&
-		    (hdr->b_l2hdr.b_daddr > taddr ||
+		/* Ensure this header has finished being written. */
+		ASSERT(!HDR_L2_WRITING(hdr));
+		ASSERT(!HDR_L2_WRITE_HEAD(hdr));
+
+		if (!all && (hdr->b_l2hdr.b_daddr >= taddr ||
```

Comment:
Yes, I did. `taddr` is the end of what we should evict (start + size). So, if a buffer starts at `taddr`, then it's beyond the area to be evicted.  That one buffer, if any, does not make a lot of difference, but just for the sake of correctness.